File: python/00234_palindrome_linked_list.py
# Python
# Time complexity - O (N)
# Space complexity - O (1)
class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True

        s = head
        f = head
        while f and f.next:
            s = s.next
            f = f.next.next

        prev = None
        current = s
        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        f = head
        s = prev
        while s:
            if f.val != s.val:
                return False
            f = f.next
            s = s.next

        return True

File: java/00234_palindrome_linked_list.java
// Java
// Time complexity - O (N)
// Space complexity - O (1)
public class PalindromeLinkedList {
    public boolean isPalindrome(ListNode head) {
        if (head == null || head.next == null) {
            return true;
        }

        ListNode s = head;
        ListNode f = head;
        while (f != null && f.next != null) {
            s = s.next;
            f = f.next.next;
        }

        ListNode prev = null;
        ListNode current = s;
        while (current != null) {
            ListNode nextNode = current.next;
            current.next = prev;
            prev = current;
            current = nextNode;
        }

        f = head;
        s = prev;
        while (s != null) {
            if (f.val != s.val) {
                return false;
            }
            f = f.next;
            s = s.next;
        }

        return true;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}
